src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -430,4 +453,35 @@ private static String[] getAdditonallyAllowedTenants() {
 
         return null;
     }
+
+    private static void addCredentialToCache(String key, TokenCredential tokenCredential) {
+        Credential credential = new Credential(tokenCredential, System.currentTimeMillis() + Credential.TIME_TO_LIVE);
+        CREDENTIAL_CACHE.put(key, credential);
+    }
+
+    private static TokenCredential getCredentialFromCache(String key) {
+        Credential credential = CREDENTIAL_CACHE.get(key);
+
+        if (null != credential) {
+            if (System.currentTimeMillis() > credential.expiryTime) {
+                CREDENTIAL_CACHE.remove(key);
+
+            } else {
+                return credential.tokenCredential;
+            }
+        }
+
+        return null;
+    }
+
+    private static class Credential {
+        TokenCredential tokenCredential;
+        long expiryTime;

```

Comment:
expiryTime doesn't seem to be used to actually expire anything. I don't think we need to expire MI or Default credential objects.